engine/resources/hetzner.go

```diff
@@ -0,0 +1,716 @@
+
+	// Full client details for interaction with the hcloud-go API.
+	client *hcloud.Client
+	// Snapshot of the server instance, for checkapply operations.
+	server *hcloud.Server
+	// Snapshot of the serverCreateOpts struct generated with hcloud.
+	serverconfig hcloud.ServerCreateOpts
+	// hcloud-formatted array of ssh keys for use with rescue mode.
+	rescuekeys []*hcloud.SSHKey
+
+	// Internal flag to simplify rescue mode operations in checkapply.
+	// False if ServerRescueMode="disabled", true otherwise.
+	rescueEnable bool
+}
+
+func (obj *HetznerRes) Default() engine.Res {
+	return &HetznerRes{
+		WaitInterval:     WaitIntervalLimit,
+		ServerRescueMode: "disabled",
+	}
+}
+
+func (obj *HetznerRes) Validate() error {
+
+	// verify name (TODO: how to ensure unique names across server instances?)
+	if obj.Name() == "" {
+		return fmt.Errorf("instance has empty name field")
+	}
+
+	// verify token (source and contents)
+	content, err := readFileLines(obj.APITokenPath)
+	if err != nil {
+		return errwrap.Wrapf(err, "no valid file in APITokenPath")
+	}
+	if content == nil {
+		return fmt.Errorf("file in APITokenPath is empty")
+	}
+	if len(content) > 1 {
+		return fmt.Errorf("file in APITokenPath contains more than one line")
+	}
+
+	// check polling interval
+	if obj.MetaParams().Poll < HetznerPollLimit {
+		return fmt.Errorf("invalid polling interval (min = %d s)", HetznerPollLimit)
+	}
+
+	// check wait interval
+	if obj.WaitInterval < WaitIntervalLimit {
+		return fmt.Errorf("invalid wait interval (min = %d)", WaitIntervalLimit)
+	}
+
+	// verify state specifiers (only allow steady states)
+	switch obj.State {
+	case StateRunning, StateOff, StateAbsent:
+		// continue, state is valid.
+	default:
+		// no other specifiers supported
+		return fmt.Errorf("invalid State")
+	}
+
+	// validate rescue mode options
+	switch obj.ServerRescueMode {
+	case ServerRescueTypeLinux32, ServerRescueTypeLinux64, ServerRescueTypeFreeBSD64:
+		// valid options for rescue mode image.
+	case ServerRescueDisabled:
+		// valid option to disable rescue mode.
+	default:
+		return fmt.Errorf("invalid ServerRescueMode")
+	}
+
+	return nil
+}
+
+func (obj *HetznerRes) Init(init *engine.Init) error {
+
+	// save init struct
+	obj.init = init
+
+	// initialize hcloud client
+	content, err := readFileLines(obj.APITokenPath)
+	token := content[0] // expecting only one!
+	if err != nil {
+		return errwrap.Wrapf(err, "error reading token in obj.APITokenPath")
+	}
+	obj.client = hcloud.NewClient(hcloud.WithToken(token))
+
+	// convert ServerRescueMode to rescueEnable flag
+	obj.rescueEnable = (obj.ServerRescueMode != ServerRescueDisabled)
+
+	return nil
+}
+
+// TODO: clear the root password info if rescue mode was enabled?
+func (obj *HetznerRes) Close() error {
+	obj.client = nil
+	obj.server = nil
+	return nil
+}
+
+func (obj *HetznerRes) Watch() error {
+	return fmt.Errorf("invalid Watch() call: Hetzner has no event API and requires polling")
```

Comment:
If you want to do it this way and require that the user use the polling metaparam, you can also check if that's setup correctly in Validate()!